src/main/java/org/javarosa/xform/parse/XFormParser.java

```diff
@@ -1841,6 +1853,32 @@ private DataBinding processStandardBindAttributes(List<String> usedAtts, List<St
         "saveIncomplete"
     ));
 
+    private String toXMLTag(Element element) {
+        String converted = "";
+        KXmlSerializer kXmlSerializer = new KXmlSerializer();
```

Comment:
Maybe it would be good to create a new class (something like `SecondaryInstanceAnalyzer`, perhaps?) to do all this. That was we could encapsulate the necessary regex, and just do all this set up once, instead of every time we get an element.

Something like (excuse my terrible Java!):

```
public class SecondaryInstanceAnalyzer {

  private KXmlSerializer;
  private StringWriter;
 
  // a list of SIs that need to be built in memory
  private inMemorySecondaryInstances;

  private INSTANCE_FUNCTION_PATTERN = Pattern.compile(/.*&$?/)
  public SecondaryInstanceAnalyzer() {
    kXmlSerializer = new KXmlSerializer();
    stringWriter = new StringWriter();
    kXmlSerializer.setOutput(stringWriter);
  }

  public analyzeElement(element) {
    // convert to text and test against regex
    // if test is positive, add to inMemorySecondaryInstances
  }

  public getInMemorySecondaryInstances() {
    return inMemorySecondaryInstances
  }

  public shouldSecondaryInstanceBeParsed(instanceId) {
    return inMemorySecondaryInstances.includes(instanceId)
  }

}
```

Then in `XFormParser` you could instantiate `SecondaryInstanceAnalyzer` and call `analyzeElement` for each element, then call `shouldSecondaryInstanceBeParsed` for each SI. If an SI is only used in a `pulldata` call, then it won't be on the list, so it won't ever get parsed.

This encapsulation is good for future extensions, i.e.

1. analyzing the XPath predicate that uses an `instance()` call to determine whether it can be replaced with SQL, which means better performance
2. adding some kind to mechanism (to be defined) to indicate to Collect that it should write a given SI to a database